### PR TITLE
Topic pool registry [9595]

### PR DESCRIPTION
**Pull request by MiguelCompany**

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2880/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2635/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2891/)

File: src/cpp/rtps/history/TopicPayloadPool.cpp [hunk 1]
```diff
@@ -278,7 +278,7 @@ bool TopicPayloadPool::shrink (
 }
 
 std::shared_ptr<ITopicPayloadPool> TopicPayloadPool::get(
-        PoolConfig config)
+        const PoolConfig& config)
```
Comment (IkerLuengo) [inline src/cpp/rtps/history/TopicPayloadPool.cpp:281]: I'm thinking we start to overuse `PoolConfig` as the input parameter (here, and as a consequence of dependencies, also in the registry). We are only using the `memory_policy` and `payload_initial_size`. Having the `initial_size` and `maximum_size` hanging around looks distracting.

Just a crazy idea: Should we have a `history_config` with all 4 values and a `pool_config` with just the first two?

**Review by IkerLuengo — COMMENTED** [review]
After some careful checking, this looks correct.

I just added some comments to discuss before approving the PR.

File: test/unittest/rtps/history/TopicPayloadPoolRegistryTests.cpp [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    EXPECT_FALSE(pool_a3);
+    EXPECT_NO_THROW(TopicPayloadPoolRegistry::release(pool_b1));
+    EXPECT_FALSE(pool_b1);
+    EXPECT_NO_THROW(TopicPayloadPoolRegistry::release(pool_b2));
+    EXPECT_FALSE(pool_b2);
+
+    // Check backups only have one reference
+    EXPECT_EQ(pool_backup_a1.use_count(), 1);
+    EXPECT_EQ(pool_backup_a3.use_count(), 1);
+    EXPECT_EQ(pool_backup_b1.use_count(), 1);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/unittest/rtps/history/TopicPayloadPoolRegistryTests.cpp:80]: done

File: src/cpp/rtps/history/TopicPayloadPool.cpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/rtps/history/TopicPayloadPool.cpp]: I've refactored `PoolConfig` to have a base `BasicPoolConfig`

**Review by IkerLuengo — APPROVED** [review]
LGTM

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2886/)
* Linux aarch64Build Status
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2641/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2897/)

**Pull request merged**